File: RunAnalyse_Simulations.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% Code to run dynamical simulations on a SpaceTime Network Model
% and visualize the results.
%
% Generates figure with:
% - timeseries, power spectra and phase order over time
% - matrices of coupling wights (C), distances (D) and Phase Coherence (PC) 
% 
% Joana Cabral January 2019 joanacabral@med.uminho.pt
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Define here the Network Spatial and Temporal structure

% Define the strenght of all existing connections between the N nodes
% as a NxN matrix 
load AAL_matrices.mat C

% Number of coupled Units
N=size(C,1); 

% Normalize such that the mean of all non-diagonal elements is 1. 
C=C/mean(C(ones(N)-eye(N)>0));

% Distance between areas
load AAL_matrices.mat D
D=D/1000; % Distance matrix in meters

% Change the order of units for visualization (optional)
Order=[1:2:N N:-2:2]; % This sorts the AAL areas into left and right hemispheres
C=C(Order,Order);
D=D(Order,Order);

%% Define here the parameters of the network model to be manipulated

    % Node natural frequency in Hz 
          f=40;    % (i.e. f=40)):          

    % Mean Delay in seconds 
          MD=0.02;  %(i.e. MD=0.01)             
       
    % Global Coupling strength
          K=5; % 
          
%% Call here the function of the Network Model

[Phases_Save, dt_save] = Kuramoto_Delays_Run(C,D,f,K,MD);


%% Process the simulated data and generate figures

figure('Name',[' K= ' num2str(K) ' and mean delay ' num2str(MD*1e3) ' ms'])

% Plot simulated time series
subplot(3,4,1:3)
plot(0:dt_save:(size(Phases_Save,2)-1)*dt_save,sin(Phases_Save'))
xlabel('Time (seconds)')
ylabel('sin(\theta)')
title(['Coupled ' num2str(f) 'Hz oscillators with for K= ' num2str(K) ' and mean delay ' num2str(MD*1e3) ' ms'])

% Power Spectrum
subplot(3,4,4)
fbins=5000;
freqZ = (0:fbins-1)/(dt_save*fbins);
ind100Hz=find(freqZ==100);
Fourier = fft(sin(Phases_Save),fbins,2);     
PSD=abs(Fourier).^2;
plot(freqZ(1:round(fbins/2)),PSD(:,1:round(fbins/2)))
xlim([0 100])
xlabel('Frequency (Hz)')
ylabel('Power')

% Plot the Order Parameter over time
subplot(3,4,5:7)
OP=abs(mean(exp(1i*Phases_Save),1));
plot(0:dt_save:(size(Phases_Save,2)-1)*dt_save,OP)
ylim([0 1])
xlabel('Time (seconds)')
ylabel('Order Parameter')

% Power Spectrum of the Order Parameter
subplot(3,4,8)
Fourier = fft(OP-mean(OP),fbins,2); %% Fourier of Z (complex)    
PSD=abs(Fourier).^2;
PSD=PSD(1:round(fbins/2))/sum(PSD(1:round(fbins/2)));
plot(freqZ(1:round(fbins/2)),PSD)
xlim([0 10])
xlabel('Frequency (Hz)')
ylabel('Power')

% Plot Connectivity and Distance Matrices and mean Phase Coherence

subplot(4,3,10)
colormap(jet)
imagesc(C,'AlphaData',~C==0)
xlabel('Nodes')
ylabel('Nodes')
title('Coupling Matrix')
axis square
colorbar

subplot(4,3,11)
imagesc(D)
xlabel('Nodes')
ylabel('Nodes')
title('Distance Matrix')
axis square
colorbar

subplot(4,3,12)
for n=1:N
    for p=1:N
        PC(:,n,p)=cos(Phases_Save(n,:)-Phases_Save(p,:));
    end
end
imagesc(squeeze(mean(PC,1)),[-1 1])
xlabel('Nodes')
ylabel('Nodes')
title('Mean Phase Coherence Matrix')
axis square
colorbar

%% Video Showing the Phase Coherence Matrix
figure
colormap jet

B0=zeros(N,1);

for t=1:size(PC,1)
    
    subplot(1,2,1)
    imagesc(squeeze(PC(t,:,:)),[-1 1])
    xlabel('Nodes')
    ylabel('Nodes')
    title(['Phase Coherence Matrix t=' num2str(t*dt_save) 's'])
    axis square

    subplot(1,2,2)
    cla
    hold on
    B1=Phases_Save(:,t);
    plot([B0 cos(B1)]',[B0 sin(B1)]','b')
    ylabel('Imag')
    xlabel('Real')
    xlim([-1 1])
    ylim([-1 1])
    axis square
    title('BOLD Phases, \Theta_n(t)')

    pause(0.01)
end

File: RunAnalyse_Simulations.py
#####################%%%%%%
#
# Code to call simulations of a network model
# and visualize the results.
#
# Generates figure with:
# - timeseries, power spectra and phase order over time
# - matrices of coupling wights, distances and 'Functional Connectivity'
#
# Joana Cabral January 2019 joanacabral@med.uminho.pt
#
#     # Adapted to Python by Laurent Perrinet Jan 2019
#
#####################%%%%%%
import numpy as np
# Define here the Network Spatial and Temporal structure

# Define the Structural Network as a NxN matrix
from scipy.io import loadmat
C = loadmat('./AAL_matrices.mat')['C']

# Number of coupled Units
N = C.shape[0]

# note: using the line below Normalizes the whole matrix by the mean
# of all non-diagonal elements is 1. 
# C=C/mean(C(ones(N)-eye(N)>0));
# here I do what you describe in the following line:
# Normalize such that the mean of all non-diagonal elements is 1.
C[np.diag(np.ones(N))==0] /= C[np.diag(np.ones(N))==0].mean()

# Distance between areas
D = loadmat('./AAL_matrices.mat')['D']
D /= 1000 # Distance matrix in meters

# Define here the parameters of the network model to be manipulated

# Node natural frequency in Hz
f = 40    # (i.e. f=40)):

# Mean Delay in seconds
MD = 0.02 # (i.e. MD = 0.01)

# Global Coupling strength
K = 5 #


# Process the simulated data and generate figures

# Call here the function of the Network Model
from Kuramoto_Delays_Run import Kuramoto_Delays_Run_AAL
Phases_Save, dt_save = Kuramoto_Delays_Run_AAL(C, D, f, K, MD)

# Process the simulated data and generate figures
import matplotlib.pyplot as plt

N_time = Phases_Save.shape[1] 
tmax = N_time / dt_save
time = np.linspace(0, tmax, N_time)

# Plot simulated time series
fig, ax = plt.subplots(1, 1, figsize=(15, 8))
ax.plot(time, np.sin(Phases_Save.T))
ax.set_xlabel('Time (seconds)')
ax.set_ylabel('sin(\theta)')
ax.set_title(['Coupled ' + str(f) + 'Hz oscillators with for K= ' + str(K) + ' and mean delay ' + str(MD*1e3) + ' ms'])
fig.suptitle(' K= ' + str(K) + ' and mean delay ' + str(MD*1e3) + ' ms')
         
# Power Spectrum
fig, axs = plt.subplots(1, 3, figsize=(15, 8))
fbins = 5000
freqZ = np.arange(fbins)/(dt_save*fbins)
ind100Hz = np.argmin(freqZ==100)
Fourier = np.fft.fft(np.sin(Phases_Save), n=fbins, axis=-1)
PSD = np.abs(Fourier)**2
ax = axs[0]
ax.plot(freqZ[:fbins//2], PSD[:, :fbins//2].mean(axis=0))
ax.set_xlim([0, 100])
ax.set_xlabel('Frequency (Hz)')
ax.set_ylabel('Power')

# Plot the Order Parameter over time
ax = axs[1]
OP = np.abs(np.mean(np.exp(1j*Phases_Save), axis=0))
ax.plot(time, OP)
ax.set_ylim([0, 1])
ax.set_xlabel('Time (seconds)')
ax.set_ylabel('Order Parameter')

# Power Spectrum of the Order Parameter
ax = axs[2]
Fourier = np.fft.fft(OP-OP.mean(), n=fbins, axis=0)
PSD = np.abs(Fourier)**2
PSD = PSD[:fbins//2]
PSD /= PSD.sum()
ax.plot(freqZ[:fbins//2], PSD)
ax.set_xlim([0, 10])
ax.set_xlabel('Frequency (Hz)')
ax.set_ylabel('Power')

# Plot Connectivity and Distance Matrices and mean Phase Coherence

# Change the order of units for visualization (optional)
Order = np.arange(0, N, 2) # This sorts the AAL areas into left and right hemispheres
Order = np.hstack((Order, np.arange(N-1, 0, -2)))

fig, axs = plt.subplots(1, 3, figsize=(15, 8))
ax = axs[0]
# colormap(jet)
ax.pcolormesh(C[Order, :][:, Order])

ax.set_xlabel('Nodes')
ax.set_ylabel('Nodes')
ax.set_title('Coupling Matrix')
ax.axis('square')
#plt.colorbar()

ax = axs[1]
ax.pcolormesh(D[Order, :][:, Order])
ax.set_xlabel('Nodes')
ax.set_ylabel('Nodes')
ax.set_title('Distance Matrix')
ax.axis('square')
#plt.colorbar()

ax = axs[2]
X = np.sin(Phases_Save)
FC = (X[:, None, :] * X[None, :, :]).mean(axis=-1)
ax.pcolormesh(FC[Order, :][:, Order])
ax.set_xlabel('Nodes')
ax.set_ylabel('Nodes')
ax.set_title('Correlation Matrix')
ax.axis('square');
#plt.colorbar()